DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/svd/src/golden_pca.hpp

```diff
@@ -0,0 +1,324 @@
+#include <math.h>
+
+#include <iomanip>
+#include <random>
+#include <fstream>
+
+/*
+This file implements the steps to
+identify principal components (Eigen vectors)
+of a matrix and finally transform input matrix
+along the directions of the principal components
+
+Following are the main steps in order to transform an
+matrix A made of multiple samples with different features.
+The matrix A will contain n samples with p features
+making it an n row, p columns matrix
+
+Here are the steps performed by this file:
+1. Generate random matrices
+2. Standardize the matrices
+3. Compute the covariance matrices of these matrices
+4. Compute Eigen vectors and Eigen values of the covariance matrices using the
+QR iteration method
+
+*/
+template <typename T>
+class GoldenPCA {
+ public:
+  int samples;                // number of samples
+  int features;               // number of features
+  int matrix_count;           // number of matrices
+  bool debug;                 // print debug information if true
+  bool benchmark_mode;        // pull data from actual dataset
+  std::vector<T> a_matrix;    // storage for input matrices
+  std::vector<T> covariance_matrix;      // storage for covariance matrices
+  std::vector<T> eigen_values;           // storage for the Eigen values
+  std::vector<T> eigen_vectors;          // storage for the Eigen vectors
+  std::vector<T> iterations;          // the number of QR iterations per matrix
+
+ private:
+  std::default_random_engine gen;
+
+ public:
+  // Constructor
+  GoldenPCA(int n, int p, int count, bool d, bool benchmark,
+            std::vector<std::vector<T> > &input_mat) {
+    samples = n;
+    features = p;
+    benchmark_mode = benchmark;
+    matrix_count = count;
+    debug = d;
+
+    a_matrix.resize(n * p * matrix_count);
+    for (int r = 0; r < n; r ++) {
+        for (int c = 0; c < p; c ++) {
+            a_matrix[r * p + c] = input_mat[r][c];
+        }
+    }
+
+    covariance_matrix.resize(p * p * matrix_count);
+
+    eigen_values.resize(p * matrix_count);
+    eigen_vectors.resize(p * p * matrix_count);
+
+    iterations.resize(matrix_count);
+  }
+
+  // Compute the covariance matrix of the matrix with index matrix_index
+  void computeCovarianceIthMatrix(int matrix_index) {
```

Comment:
```suggestion
  void ComputeCovarianceIthMatrix(int matrix_index) {
```
Variable names should be `snake_case`
Function names should be `PascalCase()`
constants should be `kCamelCase` (the leading `k` is important)